merkle/compact/tree.go

```diff
@@ -256,9 +257,9 @@ func (t *Tree) AddLeafHash(leafHash []byte, setNodeFn setNodeFunc) (int64, error
 			return assignedSeq, nil
 		}
 		// The bit is set so we have a node at that position in the nodes list so hash it with our running hash:
-		hash = t.hasher.HashChildren(t.nodes[bit], hash)
-		// Store the resulting parent hash.
-		if err := setNodeFn(bit+1, index, hash); err != nil {
+		hash := t.hasher.HashChildrenInto(t.nodes[bit], hash, hash)
+		// Store the resulting parent hash. Make a copy in case the function retains it.
+		if err := setNodeFn(bit+1, index, append(make([]byte, 0, len(hash)), hash...)); err != nil {
```

Comment:
Aside: What's the advantage of using the inline hasher here, given that we are still making copies on every iteration? Is it mostly the shared internal pre-image buffer? I wonder if a regular non-thread-safe `LogHasher` implementation that kept a similar pre-image buffer could do the same job.

Reply:
IIRC it's the sum of a shared hash buffer + not allocating a new result slice. It's pretty clear in the memory profiling results.

Reply:
Yeah, so what I am saying is that `InlineLogHasher` does eliminate allocating a new result slice, but we are still allocating one in this line (to make a copy). So the only advantage that remains is the shared hash buffer, isn't it? But for that we don't strictly need a new interface, the buffer can be hidden behind another `LogHasher` implementation.